Apply RemoveNoopLandingPads post-monomorphization

Mark-Simulacrum opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/143208)*
<!-- homu-ignore:end -->

On Cargo this cuts ~5% of the LLVM IR lines we generate (measured with -Cno-prepopulate-passes).

Closes https://github.com/rust-lang/rust/issues/159399.

Mark-Simulacrum committed:
Remove no-op cleanups on MIR post-monomorphization

This speeds up LLVM and improves codegen overall. As an example, for
cargo this cuts ~5% of the LLVM IR lines we generate (measured with
-Cno-prepopulate-passes).

Mark-Simulacrum commented:
@bors2 try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued fb9de75a813ea747c43682b22929463ada3f80de with parent 5e749eb66f93ee998145399fbdde337e57cd72ef, future [comparison URL](https://perf.rust-lang.org/compare.html?start=5e749eb66f93ee998145399fbdde337e57cd72ef&end=fb9de75a813ea747c43682b22929463ada3f80de).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.3 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (fb9de75a813ea747c43682b22929463ada3f80de): [comparison URL](https://perf.rust-lang.org/compare.html?start=5e749eb66f93ee998145399fbdde337e57cd72ef&end=fb9de75a813ea747c43682b22929463ada3f80de&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=5e749eb66f93ee998145399fbdde337e57cd72ef&end=fb9de75a813ea747c43682b22929463ada3f80de&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 18.6% | [0.2%, 60.9%]   | 6     |
| Regressions ❌ <br /> (secondary)  | 9.4%  | [0.4%, 18.4%]   | 2     |
| Improvements ✅ <br /> (primary)   | -1.9% | [-3.8%, -0.2%]  | 18    |
| Improvements ✅ <br /> (secondary) | -2.5% | [-17.6%, -0.2%] | 21    |
| All ❌✅ (primary)                 | 3.2%  | [-3.8%, 60.9%]  | 24    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=5e749eb66f93ee998145399fbdde337e57cd72ef&end=fb9de75a813ea747c43682b22929463ada3f80de&stat=max-rss)
<details>
<summary>Results (primary 8.9%, secondary 2.4%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 9.2%  | [1.3%, 33.3%]  | 34    |
| Regressions ❌ <br /> (secondary)  | 3.6%  | [2.9%, 4.9%]   | 4     |
| Improvements ✅ <br /> (primary)   | -1.6% | [-1.6%, -1.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-2.4%, -2.4%] | 1     |
| All ❌✅ (primary)                 | 8.9%  | [-1.6%, 33.3%] | 35    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=5e749eb66f93ee998145399fbdde337e57cd72ef&end=fb9de75a813ea747c43682b22929463ada3f80de&stat=cycles:u)
<details>
<summary>Results (primary 3.0%, secondary -3.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 29.8% | [26.6%, 33.1%]  | 2     |
| Regressions ❌ <br /> (secondary)  | 17.1% | [17.1%, 17.1%]  | 1     |
| Improvements ✅ <br /> (primary)   | -1.9% | [-3.0%, -1.2%]  | 11    |
| Improvements ✅ <br /> (secondary) | -6.2% | [-20.2%, -2.1%] | 8     |
| All ❌✅ (primary)                 | 3.0%  | [-3.0%, 33.1%]  | 13    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=5e749eb66f93ee998145399fbdde337e57cd72ef&end=fb9de75a813ea747c43682b22929463ada3f80de&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.7%, secondary -0.9%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-1.6%, -0.0%] | 66    |
| Improvements ✅ <br /> (secondary) | -0.9% | [-3.7%, -0.0%] | 30    |
| All ❌✅ (primary)                 | -0.7% | [-1.6%, -0.0%] | 66    |
</details>

**Bootstrap**: 695.323s -> 698.39s (0.44%)
**Artifact size**: 371.80 MiB -> 371.90 MiB (0.03%)

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued e3a4b05fb719511c7b8d0306c3a9e0445abcdfda with parent f46ce66fcc3d6058f90ac5bf0930f940f1e7b0ca, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f46ce66fcc3d6058f90ac5bf0930f940f1e7b0ca&end=e3a4b05fb719511c7b8d0306c3a9e0445abcdfda).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.8 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (e3a4b05fb719511c7b8d0306c3a9e0445abcdfda): [comparison URL](https://perf.rust-lang.org/compare.html?start=f46ce66fcc3d6058f90ac5bf0930f940f1e7b0ca&end=e3a4b05fb719511c7b8d0306c3a9e0445abcdfda&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=f46ce66fcc3d6058f90ac5bf0930f940f1e7b0ca&end=e3a4b05fb719511c7b8d0306c3a9e0445abcdfda&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.4%, 0.4%]    | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -               | 0     |
| Improvements ✅ <br /> (primary)   | -2.0% | [-4.6%, -0.2%]  | 22    |
| Improvements ✅ <br /> (secondary) | -2.8% | [-18.0%, -0.8%] | 17    |
| All ❌✅ (primary)                 | -1.9% | [-4.6%, 0.4%]   | 23    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f46ce66fcc3d6058f90ac5bf0930f940f1e7b0ca&end=e3a4b05fb719511c7b8d0306c3a9e0445abcdfda&stat=max-rss)
<details>
<summary>Results (primary -2.6%, secondary -1.3%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-3.8%, -1.2%] | 6     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-2.8%, -0.5%] | 3     |
| All ❌✅ (primary)                 | -2.6% | [-3.8%, -1.2%] | 6     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f46ce66fcc3d6058f90ac5bf0930f940f1e7b0ca&end=e3a4b05fb719511c7b8d0306c3a9e0445abcdfda&stat=cycles:u)
<details>
<summary>Results (primary -2.4%, secondary -2.8%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1%  | [1.1%, 1.1%]    | 1     |
| Regressions ❌ <br /> (secondary)  | 3.2%  | [0.8%, 7.1%]    | 4     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-4.3%, -1.2%]  | 22    |
| Improvements ✅ <br /> (secondary) | -5.0% | [-20.0%, -0.5%] | 11    |
| All ❌✅ (primary)                 | -2.4% | [-4.3%, 1.1%]   | 23    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=f46ce66fcc3d6058f90ac5bf0930f940f1e7b0ca&end=e3a4b05fb719511c7b8d0306c3a9e0445abcdfda&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.6%, secondary -0.8%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-1.4%, -0.0%] | 66    |
| Improvements ✅ <br /> (secondary) | -0.8% | [-3.6%, -0.0%] | 30    |
| All ❌✅ (primary)                 | -0.6% | [-1.4%, -0.0%] | 66    |
</details>

**Bootstrap**: 461.485s -> 461.37s (-0.02%)
**Artifact size**: 372.20 MiB -> 371.78 MiB (-0.11%)

Mark-Simulacrum commented:
@bors2 try jobs=*x86_64-msvc*

Mark-Simulacrum commented:
@bors2 try jobs=`*x86_64-msvc*`

Mark-Simulacrum commented:
@bors2 try jobs=x86_64-msvc-1,x86_64-msvc-2

bors commented:
:pushpin: Commit 676eed3ad948557fd797bf20839752671d37805c has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 676eed3ad948557fd797bf20839752671d37805c -->
<!-- homu: {"type":"Approved","sha":"676eed3ad948557fd797bf20839752671d37805c","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

cjgillot commented:
Is there an opportunity to reuse code from `RemoveNoopLandingPads` MIR pass?

RalfJung commented:
Given that comments and variable names don't line up, seems better to
@bors r-

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued d12ed3fc0cb645af2b945d13048aba82f574ee91 with parent 6677875279b560442a07a08d5119b4cd6b3c5593, future [comparison URL](https://perf.rust-lang.org/compare.html?start=6677875279b560442a07a08d5119b4cd6b3c5593&end=d12ed3fc0cb645af2b945d13048aba82f574ee91).
There are currently 14 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~14.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (d12ed3fc0cb645af2b945d13048aba82f574ee91): [comparison URL](https://perf.rust-lang.org/compare.html?start=6677875279b560442a07a08d5119b4cd6b3c5593&end=d12ed3fc0cb645af2b945d13048aba82f574ee91&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=6677875279b560442a07a08d5119b4cd6b3c5593&end=d12ed3fc0cb645af2b945d13048aba82f574ee91&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -               | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -               | 0     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-5.1%, -0.2%]  | 28    |
| Improvements ✅ <br /> (secondary) | -3.2% | [-18.1%, -0.4%] | 19    |
| All ❌✅ (primary)                 | -2.1% | [-5.1%, -0.2%]  | 28    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6677875279b560442a07a08d5119b4cd6b3c5593&end=d12ed3fc0cb645af2b945d13048aba82f574ee91&stat=max-rss)
<details>
<summary>Results (primary -2.0%, secondary 2.4%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.4%  | [2.6%, 4.2%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.4%  | [0.7%, 4.1%]   | 2     |
| Improvements ✅ <br /> (primary)   | -4.2% | [-5.6%, -3.4%] | 5     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.0% | [-5.6%, 4.2%]  | 7     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6677875279b560442a07a08d5119b4cd6b3c5593&end=d12ed3fc0cb645af2b945d13048aba82f574ee91&stat=cycles:u)
<details>
<summary>Results (primary -2.6%, secondary -4.8%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.0%  | [2.1%, 3.8%]    | 2     |
| Regressions ❌ <br /> (secondary)  | 0.7%  | [0.7%, 0.7%]    | 1     |
| Improvements ✅ <br /> (primary)   | -3.2% | [-5.6%, -1.5%]  | 18    |
| Improvements ✅ <br /> (secondary) | -5.2% | [-19.9%, -2.0%] | 14    |
| All ❌✅ (primary)                 | -2.6% | [-5.6%, 3.8%]   | 20    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=6677875279b560442a07a08d5119b4cd6b3c5593&end=d12ed3fc0cb645af2b945d13048aba82f574ee91&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.7%, secondary -1.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-1.4%, -0.0%] | 67    |
| Improvements ✅ <br /> (secondary) | -1.1% | [-3.7%, -0.0%] | 30    |
| All ❌✅ (primary)                 | -0.7% | [-1.4%, -0.0%] | 67    |
</details>

**Bootstrap**: 461.735s -> 459.425s (-0.50%)
**Artifact size**: 372.23 MiB -> 372.22 MiB (-0.00%)

Mark-Simulacrum commented:
It's not clear to me. The pre-LLVM IR we emit is just as improved in either opt or debug builds, so my best guess is that LLVM's optimization passes (SimplifyCFG, if I'm reading the code right?) are able to clean up the IR we generate quickly enough that it doesn't make an impact, and I guess in debug mode those either run later or don't run at all.

I think SimplifyCFG doesn't run at all in debug:

```
$ rustc +nightly t.rs -Zautodiff=PrintPasses
function(ee-instrument<>),always-inline,coro-cond(coro-early,cgscc(coro-split),coro-cleanup,globaldce),function(annotation-remarks),canonicalize-aliases,name-anon-globals
```

I won't copy paste the extremely big opt-level=3 pipeline, but it has simplify-cfg pretty early:

```
annotation2metadata,forceattrs,inferattrs,coro-early,function<eager-inv>(ee-instrument<>,lower-expect,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;no-switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>, ...
```

So my guess is in opt builds LLVM is partially able to clean this up early on and as a result we don't have much benefit from doing this there. But also no loss, so I'm inclined to keep it, it doesn't hurt based on our benchmarks and there are a few opt binary size wins (https://perf.rust-lang.org/compare.html?start=6677875279b560442a07a08d5119b4cd6b3c5593&end=d12ed3fc0cb645af2b945d13048aba82f574ee91&stat=size%3Alinked_artifact&debug=false).

klensy commented:
> rustc +nightly t.rs -Zautodiff=PrintPasses

Weird, why this works

```
    -Z                                      autodiff=val -- a list of autodiff flags to enable
        Mandatory setting:
        `=Enable`

```

https://github.com/rust-lang/rust/blob/556d20a834126d2d0ac20743b9792b8474d6d03c/compiler/rustc_codegen_llvm/src/builder/autodiff.rs#L473-L480

Mark-Simulacrum commented:
I think that could make sense but I'd see it as potentially separate -- this change feels non-invasive to me (e.g., not hurting debuginfo quality) whereas SimplifyCFG seems like it could do more than that. I'm also not sure how much we want to get into changing the LLVM pass infrastructure...

bors commented:
:hourglass: Testing commit ec26ddedd9e95309115f0adc575aaa8405bae2bd with merge 56e872c04ec4d0aede16d304f98d91ec05c1ae2a...
<!-- homu: {"type":"BuildStarted","head_sha":"ec26ddedd9e95309115f0adc575aaa8405bae2bd","merge_sha":"56e872c04ec4d0aede16d304f98d91ec05c1ae2a"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/16130229816/job/45517230988)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/16130229816/job/45517230988","builder_name":"checks-actions"} -->

jdonszelmann commented:
@bors r- seems to have failed ci already, unsure why r+

oli-obk reviewed (approved):
@bors r+

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/143208/b26c8ef0535b2de24a0af4048370caf449eebabd..f513294c00bd45c666381cc381831c2bcd3e674b)*

Mark-Simulacrum committed:
Adjust line-tables-only on i686-msvc to ignore bar as well

My suspicion is that it's getting inlined; the functions have
`#[no_mangle]` but do not have `#[inline(never)]`. `#[no_mangle]` is
already iffy on generics.

I don't think it's worth trying to figure out why foo and bar are
special on 32-bit msvc.
